src/marketplace.js

```diff
@@ -0,0 +1,321 @@
+                    await saveFile(this.rootKeyPairPath, this.rootKeyPair.toWIF());
+                }
+                source = 'random';
+            }
+            const buffer = this.rootKeyPair.getPublicKeyBuffer();
+            this.pubKeyMultihash = await this.chluIpfs.instance.vendor.storePublicKey(buffer);
+            this.chluIpfs.pin(this.pubKeyMultihash);
+            // TODO: request pin?
+        }
+        return {
+            keyPair: this.rootKeyPair,
+            pubKeyMultihash: this.pubKeyMultihash,
+            source
+        }; 
+    }
+
+    /**
+     * Get the list of vendors that is registered in this
+     * marketplace. The ID used to identify each vendor
+     * is the multihash of their public key, which can be
+     * used to get the pub key from IPFS.
+     * 
+     * @returns {Array.<string>}
+     * @memberof Marketplace
+     */
+    async getVendorIDs() {
+        await this.start();
+        return await this.db.getVendorIDs();
+    }
+
+    /**
+     * @typedef {Object} Vendor
+     * @property {string} vPubKeyMultihash the multihash of the vendor key (aka Vendor ID)
+     * @property {string} vmPubKeyMultihash the multihash of the vendor-marketplace key
+     * @property {string} mSignature hex encoded signature of the marketplace for the vendor-marketplace key
+     * @property {string} vSignature hex encoded signature of the vendor for the vendor-marketplace key
+     */
+
+    /**
+     * Get public data about a specific vendor
+     * 
+     * @param {string} id the multihash of this vendor's pub key, aka their ID
+     * @returns {Promise<Vendor>}
+     * @memberof Marketplace
+     */
+    async getVendor(id) {
+        await this.start();
+        const v = await this.db.getVendor(id);
+        if (v) {
+            // the full keypair is omitted intentionally
+            // so it does not get out by mistake
+            return {
+                mSignature: v.mSignature,
+                vPubKeyMultihash: v.vPubKeyMultihash,
+                vSignature: v.vSignature,
+                vmPubKeyMultihash: v.vmPubKeyMultihash
+            };
+        } else {
+            throw new Error('Vendor with key ' + id + ' is not registered');
+        }
+    }
+
+    /**
+     * Get the vendor-marketplace key pair for a vendor
+     * 
+     * @param {string} vendorId the vendor's public key multihash
+     * @returns {Promise<ECPair>}
+     * @memberof Marketplace
+     */
+    async getVMKeyPair(vendorId) {
+        await this.start();
+        const wif = await this.db.getVMKeyPairWIF(vendorId);
+        return ECPair.fromWIF(wif);
+    }
+
+    async registerVendor(vendorPubKeyMultihash) {
+        const id = vendorPubKeyMultihash;
+        await this.start();
+        const vmKeyPair = ECPair.makeRandom();
+        const pubKeyBuffer = vmKeyPair.getPublicKeyBuffer();
+        const vmPubKeyMultihash = await this.chluIpfs.instance.vendor.storePublicKey(pubKeyBuffer);
```

Comment:
catch error, return meaningful error and make sure we rollback any persistent changes made up to this point.